presto-common/src/main/java/com/facebook/presto/common/zorder/ZOrderIntegers.java

```diff
@@ -0,0 +1,167 @@
+            throw new IllegalArgumentException(format("Our current Z-Ordering implementation does not support more than %d input numbers.", MAX_INPUT_DIMENSIONS));
+        }
+
+        if (input.size() > encodingBits.size()) {
+            throw new IllegalArgumentException(format("Input list size (%d) does not match encoding bits list size (%d).",
+                    input.size(), encodingBits.size()));
+        }
+
+        for (int index = 0; index < input.size(); index++) {
+            if (input.get(index) < 0) {
+                throw new IllegalArgumentException("Our current Z-Ordering implementation does not support negative input numbers.");
+            }
+        }
+    }
+
+    private void checkEncodeInputValidity(List<Integer> input, AddressType type)
+            throws IllegalArgumentException
+    {
+        checkEncodeInputValidity(input);
+
+        int maximumBits;
+        switch (type) {
+            case INT:
+                maximumBits = 32;
+                break;
+            case LONG:
+                maximumBits = 64;
+                break;
+            default:
+                throw new IllegalArgumentException("Our current Z-Ordering implementation only supports int and long address types.");
+        }
+
+        if (totalBitLength > maximumBits) {
+            throw new IllegalArgumentException(format("The z-address type specified is not large enough to hold %d bits.", totalBitLength));
+        }
+    }
+
+    private long convertByteAddressToLong(byte[] byteAddress)
+    {
+        long address = 0;
+        for (int byteIndex = 0; byteIndex < byteAddress.length; byteIndex++) {
+            address |= Byte.toUnsignedLong(byteAddress[byteIndex]) << (byteIndex << 3);
+        }
+        return address;
+    }
+
+    @Override
+    public byte[] encodeToByteArray(List input)
+            throws IllegalArgumentException
+    {
+        checkEncodeInputValidity(input);
+
+        byte[] address = new byte[(totalBitLength + 7) >> 3];
+        int bitIndex = totalBitLength - 1;
+        // Interweave input bits into address from the most significant bit to preserve data locality
+        for (int bitsProcessed = 1; bitsProcessed <= maxBitLength; bitsProcessed++) {
+            for (int index = 0; index < input.size(); index++) {
+                if (bitsProcessed > encodingBits.get(index)) {
+                    continue;
+                }
+                int bitPosition = encodingBits.get(index) - bitsProcessed;
+                byte maskedBit = (byte) (((int) input.get(index) >> bitPosition) & 1);
+                address[bitIndex >> 3] |= maskedBit << (bitIndex & 7);
+                bitIndex--;
+            }
+        }
+        return address;
+    }
+
+    @Override
+    public long encodeToNumber(List input, AddressType type)
+    {
+        checkEncodeInputValidity(input, type);
+
+        return convertByteAddressToLong(encodeToByteArray(input));
+    }
```

Comment:
Do we need int version?

Reply:
The return type is long, but casting to int can be automatic. If the user chooses AddressType.INT, the input validity check ensures that the address won't be longer than 32 bits. So I don't think we need an int version of the function.